mlir/lib/Quantum/Transforms/MergeRotationsPatterns.cpp

```diff
@@ -0,0 +1,131 @@
+// Copyright 2024 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#define DEBUG_TYPE "merge-rotations"
+
+#include "Quantum/IR/QuantumOps.h"
+#include "Quantum/Transforms/Patterns.h"
+#include "VerifyParentGateAnalysis.hpp"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "llvm/ADT/StringSet.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/Errc.h"
+
+using llvm::dbgs;
+using namespace mlir;
+using namespace catalyst::quantum;
+
+static const mlir::StringSet<> rotationsSet = {"RX",  "RY",  "RZ",  "PhaseShift",           "Rot",
+                                               "CRX", "CRY", "CRZ", "ControlledPhaseShift", "CRot"};
+
+namespace {
+
+struct MergeRotationsRewritePattern : public mlir::OpRewritePattern<CustomOp> {
+    using mlir::OpRewritePattern<CustomOp>::OpRewritePattern;
+
+    mlir::LogicalResult matchAndRewrite(CustomOp op, mlir::PatternRewriter &rewriter) const override
+    {
+        LLVM_DEBUG(dbgs() << "Simplifying the following operation:\n" << op << "\n");
+        auto loc = op.getLoc();
+        StringRef opGateName = op.getGateName();
+
+        if (!rotationsSet.contains(opGateName))
+            return failure();
+        ValueRange inQubits = op.getInQubits();
+        auto parentOp = dyn_cast_or_null<CustomOp>(inQubits[0].getDefiningOp());
+
+        VerifyParentGateAndNameAnalysis<CustomOp> vpga(op);
```

Comment:
```suggestion
        VerifyParentGateAndNameAnalysis vpga(op);
```

since the named analysis is now explicitly on customop and is no longer templated